itest/lnd_channel_backup_test.go

```diff
@@ -1673,3 +1954,389 @@ func assertDLPExecuted(ht *lntest.HarnessTest,
+
+		// Mine one block to trigger the sweeper to sweep.
+		ht.MineEmptyBlocks(1)
+
+		// Dave's sweeper swept the outputs of force close tx.
+		if lntest.CommitTypeHasAnchors(commitType) {
+			ht.AssertNumPendingSweeps(dave, 2)
+		} else {
+			ht.AssertNumPendingSweeps(dave, 1)
+		}
+
+		// Assert the sweeping tx is mined.
+		ht.MineBlocksAndAssertNumTxes(1, 1)
+
+		// Funds are now in Dave's wallet.
+		ht.AssertNumPendingForceClose(dave, 0)
+
+		// Now the channel should be fully closed also from
+		// Carol's POV.
+		ht.AssertNumPendingForceClose(carol, 0)
+	}
+
+	// We query Dave's balance to make sure it increased after the channel
+	// closed. This checks that he was able to sweep the funds he had in
+	// the channel.
+	daveBalResp := dave.RPC.WalletBalance()
+	daveBalance := daveBalResp.ConfirmedBalance
+	require.Greater(ht, daveBalance, daveStartingBalance,
+		"balance not increased")
+
+	ht.AssertNodeNumChannels(dave, 0)
+
+	// Make sure Carol got her balance back.
+	err := wait.NoError(func() error {
+		carolBalResp := carol.RPC.WalletBalance()
+		carolBalance := carolBalResp.ConfirmedBalance
+
+		// With Neutrino we don't get a backend error when trying to
+		// publish an orphan TX (which is what the sweep for the remote
+		// anchor is since the remote commitment TX was not broadcast).
+		// That's why the wallet still sees that as unconfirmed and we
+		// need to count the total balance instead of the confirmed.
+		if ht.IsNeutrinoBackend() {
+			carolBalance = carolBalResp.TotalBalance
+		}
+
+		if carolBalance <= carolStartingBalance {
+			return fmt.Errorf("expected carol to have balance "+
+				"above %d, instead had %v",
+				carolStartingBalance, carolBalance)
+		}
+
+		return nil
+	}, defaultTimeout)
+	require.NoError(ht, err, "timeout while checking carol's balance")
+
+	ht.AssertNodeNumChannels(carol, 0)
+}
+
+// assertPenalty asserts that Carol penalized Dave for using an outdated
+// force close transaction.
+func assertPenalty(ht *lntest.HarnessTest,
+	carol *node.HarnessNode, carolStartingBalance int64,
+	dave *node.HarnessNode, daveStartingBalance int64,
+	startCarol func() error) {
+
+	ht.Helper()
+
+	// Detect manually broadcasted Dave's force close tx.
+	ht.Miner.AssertNumTxsInMempool(1)
+
+	// Dave should consider the channel "waiting close", as he noticed the
+	// channel was out of sync, and is now waiting for a force close to hit
+	// the chain.
+	ht.AssertNumWaitingClose(dave, 1)
+
+	// Restart Dave to make sure he is able to sweep the funds after
+	// shutdown.
+	ht.RestartNode(dave)
+
+	// Generate a single block, which should confirm the closing tx.
+	ht.MineBlocksAndAssertNumTxes(1, 1)
+
+	// Start Carol after mining closing tx if requested. It is done here to
+	// ensure that she does not publish her version of force close tx.
+	require.NoError(ht, startCarol())
+	ht.EnsureConnected(carol, dave)
+
+	// Wait for Carol to publish a penalty transaction.
+	require.Eventually(ht, func() bool {
```

Comment:
I think `MineBlocksAndAssertNumTxes` already waits the tx?

Reply:
Great catch!

I added `require.Eventually` because I got a failure in CI in `ht.MineBlocksAndAssertNumTxes(1, 1)` call (it got 0 txs), I [wrote a comment about it](https://github.com/lightningnetwork/lnd/pull/8183#issuecomment-2232295216) and thought mistakenly that `MineBlocksAndAssertNumTxes` doesn't wait.

I checked the code of `MineBlocksAndAssertNumTxes` and `AssertNumTxsInMempool` and now I see that it was not the reason of that failure, because `AssertNumTxsInMempool` waits for 1 minute. I don't know why there were no transactions in mempool in that run of the test. Penalty transaction has to be there... I'll try to undo `require.Eventually` and reproduce it locally and debug.

Reply:
I rebased the PR and fixed itest build errors (changes in ht interface and chanRestoreViaRPC function).

Also removed waiting for tx in mempool and added waiting for ClosedChannels call next to it, since it failed sometimes, because a channel was still closing during the call (so it saw 0 closed channels).

However I can't reproduce the original failure locally. For some reason Carol node didn't issue a justice transaction in that run in CI. Do you have any ideas why it could happen?